consult.el

```diff
@@ -3010,48 +3014,74 @@ INITIAL is the initial input."
 	     (unless (or (bolp) (eobp))
 	       (forward-line 0))
 	     (point))))
-        (consult--each-line beg end
-          (push (buffer-substring-no-properties beg end) lines)
-          (push (make-overlay beg (1+ end)) overlays))))
-    (unless (use-region-p)
-      (goto-char (point-min)))
+        (let ((i 0) line)
+	  (consult--each-line beg end
+	    (setq line
+		  (if (eq beg end) (char-to-string ?\n) ; "\n" reuses string! 
+		    (buffer-substring-no-properties beg end)))
+	    (add-text-properties 0 1 `(line (,(cl-incf i) ,beg ,end)) line)
+	    (push line lines))
+	  (setq lines (nreverse lines)))))
     (lambda (input restore)
       ;; New input provided -> Update
       (when (and input (not (equal input last-input)))
-        (if (string-match-p "\\`!? ?\\'" input)
-            ;; Special case the empty input for performance.
-            (progn
-              (dolist (ov overlays)
-                (overlay-put ov 'invisible nil))
-              (setq last-input input))
+	(mapc #'delete-overlay overlays)
+	(setq last-input input overlays nil)
+        (unless (string-match-p "\\`!? ?\\'" input) ; empty input.
           (let* ((not (string-prefix-p "! " input))
                  (stripped (string-remove-prefix "! " input))
                  ;; Heavy computation is interruptible if *not* committing!
-                 (ht (if restore
-                         (consult--string-hash (funcall filter stripped lines))
-                       (while-no-input
-                         (consult--string-hash (funcall filter stripped lines))))))
-            (when (hash-table-p ht)
-              (let ((ov overlays) (li lines))
-                (while ov
-                  (overlay-put (car ov) 'invisible (eq not (gethash (car li) ht)))
-                  (setq li (cdr li) ov (cdr ov))))
-              (setq last-input input)))))
+                 (matches (if restore
+			      (consult--extract-line-ind
```

Comment:
consult--extract-line-ind should be inlined. Not here but later in the while loop. We should avoid allocating a new list with the mapcar.